src/layers/ImageLayer.js

```diff
@@ -77,13 +79,36 @@ export default class ImageLayer extends CompositeLayer {
     if (loaderChanged || loaderSelectionChanged) {
       // Only fetch new data to render if loader has changed
       const { loader, z, loaderSelection } = this.props;
-      loader.getRaster({ z, loaderSelection }).then(({ data, width, height }) =>
-        this.setState({
-          data: !isWebGL2(this.context.gl) ? to32BitFloat(data) : data,
-          height,
-          width
-        })
-      );
+      loader.getRaster({ z, loaderSelection }).then(imageData => {
+        const isInterleavedAndRGB = loader.isInterleaved && loader.isRgb;
+        // eslint-disable-next-line no-param-reassign
+        const data =
+          // eslint-disable-next-line no-nested-ternary
+          isInterleavedAndRGB
+            ? imageData.data[0]
+            : !isWebGL2(this.context.gl)
+            ? to32BitFloat(imageData.data)
+            : imageData.data;
+        const { height, width } = imageData;
+        if (
+          data.length === imageData.width * imageData.height * 3 &&
+          isInterleavedAndRGB
+        ) {
+          this.setState({
+            data,
+            height,
+            width,
+            format: GL.RGB,
+            dataFormat: GL.RGB
+          });
+        } else {
+          this.setState({
+            data,
+            height,
+            width
+          });
+        }
+      });
```

Comment:
I'm having a really hard time reading this, especially with the ternaries. It's hard to tell what ends up as what. I'd probably just mutate the `imageData` params and have explicit blocks:

```javascript
      loader.getRaster({ z, loaderSelection }).then(raster => {
       if(loader.isInterleaved && loader.isRgb) {
          // data is for BitmapLayer and needs to be of form { data: Uint8Array, width, height };
          raster.data = raster.data[0];
          if (raster.data.length === raster.width * raster.height) {
            // data is RGB (not RGBA) and need to update texture formats
            raster.format = GL.RGB;
            raster.dataFormat = GL.RGB;
          }
       } else if (!isWebGL2(this.context.gl)) {
         // data is for XLRLayer in non-WebGL2 evironment
         // we need to convert data to compatible textures 
         raster.data = to32BitFloat(raster.data);
       }
       this.setState(raster);
      });

```

Reply:
- block 1: data is RGB, need to get first typedarray and check if we need to update GL format
- block 2: data is for XRLayer but no web-gl, need to cast data

There is a no-op to `raster` if the raster isn't RGB and WebGL2 is supported.

Reply:
This is much clearer, thanks.